lib/puppet/file_serving/metadata.rb

```diff
@@ -88,6 +90,21 @@ def collect_stat(path)
   def collect(source_permissions = nil)
     real_path = full_path
 
+    if static_catalog
+      if environment && environment.static_catalogs?
+        env_path = File.join(Puppet[:environmentpath], environment.name.to_s) # environment.name is a symbol sometimes
+        puts "Checking if #{real_path} is in #{env_path}"
+        if real_path.start_with? env_path
```

Comment:
I actually explicitly wanted to return a metadata object even when it's not inlinable - this lets us differentiate between the "the file is not inlinable" case and the "the file does not exist" case, so we can throw a compilation error on the latter. I'd be fine still returning the full metadata with an "is this inlinable?" attribute or method, though.

I still kind of prefer to do the actual inlinable determination in the `collect` call, since in the symlink follow case that involves additional stats, and I want to make it clear where "expensive" work happens

Re a couple of your other concerns:
* It looks like `path2instances` just calls out to the singular version, passing along the same request object. Options should make it just fine?
* The agent calling the same code is awful, but all the new behavior is gated on the `static_catalog` option being set, which it only will be for requests coming from the compiler